### ISSUE TITLE
Unexpected compile error on union types

### ISSUE THREAD

**Issue by mysticatea** [Microsoft/TypeScript#22181]
**TypeScript Version:**  2.8.0-dev.20180224

**Search Terms:** `union type`, `literal type`

**Code**

```ts
interface A {
    type: "a"
}
interface B {
    type: "b"
}
type AorB = A | B

function test1(type: AorB["type"]): AorB {
    return { type }
}
function test2(type: "a" | "b"): AorB {
    return { type }
}
```

**Expected behavior:**

No errors because the parameter `type` is a value of `AorB["type"]` so `{ type }` should be an `AorB`.

**Actual behavior:**

Errors exist on both test1 and test2.

```
Type '{ type: "a" | "b"; }' is not assignable to type 'AorB'.
  Type '{ type: "a" | "b"; }' is not assignable to type 'B'.
    Types of property 'type' are incompatible.
      Type '"a" | "b"' is not assignable to type '"b"'.
        Type '"a"' is not assignable to type '"b"'.
```

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=interface%20A%20%7B%0D%0A%20%20%20%20type%3A%20%22a%22%0D%0A%7D%0D%0Ainterface%20B%20%7B%0D%0A%20%20%20%20type%3A%20%22b%22%0D%0A%7D%0D%0Atype%20AorB%20%3D%20A%20%7C%20B%0D%0A%0D%0Afunction%20test1(type%3A%20AorB%5B%22type%22%5D)%3A%20AorB%20%7B%0D%0A%20%20%20%20return%20%7B%20type%20%7D%0D%0A%7D%0D%0Afunction%20test2(type%3A%20%22a%22%20%7C%20%22b%22)%3A%20AorB%20%7B%0D%0A%20%20%20%20return%20%7B%20type%20%7D%0D%0A%7D

**Related Issues:**

- #21297 might be?

**Comment by HerringtonDarkholme** [comment]
Generally it will lose information/generate unsafe code when collapsing two union types into one. 
Consider: 

```ts
interface A { type: 'A', value?: number }
interface B { type: 'B', value?: string }

type AorB = A | B
function test2(type: "a" | "b"): AorB {
    return {type, value: 123} // wrong, when type === 'b'
}
```

The correct checking should performed as checking if the source type is assignable to  any constituent  type of union type. 

One might suggest if two constituent types have only one discriminate property, we can merge two interfaces. However, TypeScript also allows interface augmentation where users can add properties to interface later. I suspect the above merging strategy will benefit.

**Comment by sylvanaar** [comment]
This one seems to come up all the time. Maybe the error message can be improved?

`AorB` means `{ type: "a" } | { type: "b" }` not `{ type: "a" | "b" }`
